// @brojs/fire.app/src/virtual-module.ts
import history from './history';

const createVirtualModule = ({ config, navigations, features }: {
    config: {
        [key: string]: string
    },
    navigations: {
        [key: string]: string
    },
    features: {
        [key: string]: {
            [key: string]: string
        }
    }

}) => ({
    getConfig: () => config,
    getConfigValue: (key: string) => config[key],
    getNavigations: () => navigations,
    getNavigation: () => navigations,
    getNavigationsValue: (pkg: string) => navigations[pkg],
    getNavigationValue: (pkg: string) => navigations[pkg],
    getAllFeatures: () => features,
    getFeatures: (pkg: string) => features[pkg],
    getFeatureValue: <ReturnValue = string>(pkg: string, featureName: string, defaultValue?: ReturnValue) => features[pkg]?.[featureName] as ReturnValue ?? defaultValue,
    getHistory: () => history,
})

export const defineVirtualModule = (params) => {
    const virtualModule = createVirtualModule(params)
    System.set('https://brojs.ru/virtual-module.js', {
        ...virtualModule
    });
}
